// @nacelle/storefront-sdk/src/types/parameters.ts
import { NavigationFilterInput } from '../types';

export interface FetchMethodParams {
  nacelleEntryIds?: string[];
  handles?: string[];
  locale?: string;
  maxReturnedEntries?: number;
  startAfterEntryId?: string;
  cursor?: string;
  edgesToNodes?: boolean;
  advancedOptions?: FetchMethodAdvancedParams;
}

export interface FetchMethodAdvancedParams {
  // For adjusting how the SDK internally paginates. Not needed for most
  // use cases.
  entriesPerPage?: number;
}

export interface FetchCollectionMethodParams extends FetchMethodParams {
  // For .productCollections and .contentCollections this limits the number
  // of entries for each collection
  maxReturnedEntriesPerCollection?: number;
}

export interface FetchContentMethodParams extends FetchMethodParams {
  type?: string;
}

export interface QueryParams {
  query: string;
  variables?: string;
}

export type NavigationParams = NavigationFilterInput;

export interface FetchCollectionEntriesMethodParams {
  collectionEntryId?: string;
  handle?: string;
  locale?: string;
  maxReturnedEntries?: number;
  startAfterEntryId?: string;
  cursor?: string;
  edgesToNodes?: boolean;
  advancedOptions?: FetchMethodAdvancedParams;
}
