// @nacelle/storefront-sdk/src/graphql/queries/content-collections-query.ts
import contentCollectionFragment from '../fragments/content-collection';
import contentFragment from '../fragments/content';
import collectionContentFragment from '../fragments/collection-content';

const contentCollectionsQuery = `
query contentCollectionsQuery(
  $filter: ContentCollectionFilterInput,
  $maxReturnedEntriesPerCollection: Int
) {
  contentCollections(filter: $filter) {
    ${contentCollectionFragment}
    entries(first: $maxReturnedEntriesPerCollection) {
      ${contentFragment}
    }
    content {
      ${collectionContentFragment}
    }
  }
}
`;

const contentCollectionEntriesQuery = `
query contentCollectionEntries(
  $filter: ContentCollectionFilterInput,
  $contentFirst: Int,
  $contentAfter: String
) {
  contentCollections(filter: $filter) {
    entries(first: $contentFirst, after: $contentAfter) {
      ${contentFragment}
    }
  }
}
`;

const allContentCollectionsQuery = `
query allContentCollectionsQuery(
  $filter: ContentCollectionFilterInput,
  $maxReturnedEntriesPerCollection: Int
) {
  allContentCollections(filter: $filter) {
    pageInfo {
      hasNextPage
    }
    edges {
      cursor
      node {
        ${contentCollectionFragment}
        contentConnection(first: $maxReturnedEntriesPerCollection) {
          pageInfo {
            hasNextPage
          }
          edges {
            cursor
            node {
              ${contentFragment}
            }
          }
        }
        content {
          ${collectionContentFragment}
        }
      }
    }
  }
}
`;

const allContentCollectionEntriesQuery = `
query contentCollectionEntries(
  $filter: ContentCollectionFilterInput,
  $entriesFirst: Int,
  $entriesAfter: String
) {
  allContentCollections(filter: $filter) {
    edges {
      node {
        contentConnection(first: $entriesFirst, after: $entriesAfter) {
          pageInfo {
            hasNextPage
          }
          edges {
            cursor
            node {
              ${contentFragment}
            }
          }
        }
      }
    }
  }
}
`;

export {
  contentCollectionsQuery,
  contentCollectionEntriesQuery,
  allContentCollectionsQuery,
  allContentCollectionEntriesQuery,
};
