// @nacelle/storefront-sdk/src/graphql/queries/product-collections-query.ts
import { productCollectionFragmentSubset } from '../fragments/product-collection';
import productFragment from '../fragments/product';
import collectionContentFragment from '../fragments/collection-content';

const productCollectionsQuery = `
query productCollections(
  $filter: ProductCollectionFilterInput,
  $maxReturnedEntriesPerCollection: Int
) {
  productCollections(filter: $filter) {
    ${productCollectionFragmentSubset}
    content {
      ${collectionContentFragment}
    }
    products(first: $maxReturnedEntriesPerCollection) {
      ${productFragment}
    }
  }
}
`;

const productCollectionEntriesQuery = `
query productCollectionEntries(
  $filter: ProductCollectionFilterInput,
  $entriesFirst: Int,
  $entriesAfter: String
) {
  productCollections(filter: $filter) {
    products(first: $entriesFirst, after: $entriesAfter) {
      ${productFragment}
    }
  }
}
`;

const allProductCollectionsQuery = `
query allProductCollections(
  $filter: ProductCollectionFilterInput,
  $maxReturnedEntriesPerCollection: Int
) {
  allProductCollections(filter: $filter) {
    pageInfo {
      hasNextPage
    }
    edges {
      cursor
      node {
        ${productCollectionFragmentSubset}
        content {
          ${collectionContentFragment}
        }
        productConnection(first: $maxReturnedEntriesPerCollection) {
          pageInfo {
            hasNextPage
          }
          edges {
            cursor
            node {
              ${productFragment}
            }
          }
        }
      }
    }
  }
}
`;

const allProductCollectionEntriesQuery = `
query productCollectionEntries(
  $filter: ProductCollectionFilterInput,
  $entriesFirst: Int,
  $entriesAfter: String
) {
  allProductCollections(filter: $filter) {
    edges {
      node {
        productConnection(first: $entriesFirst, after: $entriesAfter) {
          pageInfo {
            hasNextPage
          }
          edges {
            cursor
            node {
              ${productFragment}
            }
          }
        }
      }
    }
  }
}
`;

export {
  productCollectionsQuery,
  productCollectionEntriesQuery,
  allProductCollectionsQuery,
  allProductCollectionEntriesQuery,
};
